relational_operators/SampleOperator.hpp

```diff
@@ -0,0 +1,177 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_RELATIONAL_OPERATORS_SAMPLE_OPERATOR_HPP_
+#define QUICKSTEP_RELATIONAL_OPERATORS_SAMPLE_OPERATOR_HPP_
+
+#include <memory>
+#include <vector>
+
+#include "catalog/CatalogRelation.hpp"
+#include "catalog/CatalogTypedefs.hpp"
+#include "query_execution/QueryContext.hpp"
+#include "relational_operators/RelationalOperator.hpp"
+#include "relational_operators/WorkOrder.hpp"
+#include "storage/StorageBlockInfo.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+#include "tmb/id_typedefs.h"
+
+namespace tmb { class MessageBus; }
+
+namespace quickstep {
+
+class CatalogDatabase;
+class CatalogRelationSchema;
+class InsertDestination;
+class Predicate;
+class Scalar;
+class StorageManager;
+class WorkOrdersContainer;
+
+
+/** \addtogroup RelationalOperators
+ *  @{
+ */
+
+/**
+ * @brief An operator which performs sample (block or tuple) over a relation.
+ **/
+class SampleOperator : public RelationalOperator {
+ public:
+  /**
+   * @brief Constructor for Sample with the sampling percentage and type of samplng
+   *
+   * @param input_relation The relation to perform sampling over.
+   * @param output_relation The output relation.
+   * @param output_destination_index The index of the InsertDestination in the
+   *        QueryContext to insert the sampling results.
+   * @param input_relation_is_stored If input_relation is a stored relation and
+   *        is fully available to the operator before it can start generating
+   *        workorders.
+   * @param is_block_sample Flag indicating whether the sample type is block or tuple
+   * @param percentage The percentage of data to be sampled
+   * 
+   **/
+  SampleOperator(const CatalogRelation &input_relation,
+                 const CatalogRelation &output_relation,
```

Comment:
Sure.  If I make the input CatalogRelationSchema , how do i get the block ids ?

Previously this was the code:
input_relation_block_ids_(input_relation_is_stored ? input_relation.getBlocksSnapshot()
                                                       : std::vector<block_id>()),